### PR TITLE
Make it possible to disable strict parsing in email module

### PR DESCRIPTION
**Pull request by frenzymadness**
I am creating a PR for easier review. This single commit is meant to be applied in RHEL after the patch 415 with the actual fix from the upstream.

This is a solution we (with @encukou) came up with.

Documentation change TBD when the implementation is approved.

### CODE REVIEW

**Comment by frenzymadness** [comment]
Cc: @vstinner

File: Lib/email/utils.py [hunk 1]
```diff
@@ -48,6 +48,45 @@
...
+            return _cached_strict_addr_parsing
+
+        try:
+            file = open(_EMAIL_CONFIG_FILE)
+        except FileNotFoundError:
+            pass
+        else:
+            with file:
+                import configparser
+                config = configparser.ConfigParser(
```
Comment (vstinner) [inline Lib/email/utils.py]: I wanted to suggest SafeConfigParser, but `interpolation=None` does the same, so it's fine.

File: Lib/email/utils.py [hunk 2]
```diff
@@ -48,6 +48,45 @@
...
+                    interpolation=None,
+                    comment_prefixes=('#', ),
+
+                )
+                config.read_file(file)
+                disabled = config.getboolean('email_addr_parsing', "PYTHON_EMAIL_DISABLE_STRICT_ADDR_PARSING", fallback=None)
```
Comment (vstinner) [inline Lib/email/utils.py]:
```suggestion
                disabled = config.getboolean('email_addr_parsing', "PYTHON_EMAIL_DISABLE_STRICT_ADDR_PARSING", fallback=False)
```

File: Lib/email/utils.py [hunk 3]
```diff
@@ -48,6 +48,45 @@
...
+    """
+    global _cached_strict_addr_parsing
+
+    try:
+        return _cached_strict_addr_parsing
+    except NameError:
+        disabled = bool(os.environ.get("PYTHON_EMAIL_DISABLE_STRICT_ADDR_PARSING"))
+        if disabled:
+            _cached_strict_addr_parsing = False
+            return _cached_strict_addr_parsing
```
Comment (vstinner) [inline Lib/email/utils.py]: You may move the code into a new `_use_strict_email_parsing_impl()` function, and only keep the code managing the cache here. Something like:

```py
_cached_strict_addr_parsing = None

def _use_strict_email_parsing():
    """"...doc..."""
    global _cached_strict_addr_parsing
    if _use_strict_email_parsing is None:
        _use_strict_email_parsing = _use_strict_email_parsing_impl()
    return _use_strict_email_parsing
```

File: Lib/test/test_email/test_email.py [hunk 4]
```diff
@@ -3313,6 +3316,78 @@ def test_parsing_errors(self):
         # Test email.utils.supports_strict_parsing attribute
         self.assertEqual(email.utils.supports_strict_parsing, True)
 
+    def test_parsing_errors_strict_disabled_via_env_var(self):
+        address = 'alice@example.org )Alice('
+        empty = ('', '')
```
Comment (vstinner) [inline Lib/test/test_email/test_email.py:3321]: For the test, you can just call _use_strict_email_parsing() rather than testing email parsing.

File: Lib/test/test_email/test_email.py [hunk 5]
```diff
@@ -3313,6 +3316,78 @@ def test_parsing_errors(self):
...
+        # Delete cached default value to make the function
+        # reload the environment variable provided below.
+        try:
+            del utils._cached_strict_addr_parsing
+        except AttributeError:
+            pass
+
+        # Strict disabled via env variable, old behavior expected
+        with EnvironmentVarGuard() as environ:
+            environ["PYTHON_EMAIL_DISABLE_STRICT_ADDR_PARSING"] = "1"
```
Comment (vstinner) [inline Lib/test/test_email/test_email.py:3329]: Can you add an test where the env var is set to an empty string?

File: Lib/test/test_email/test_email.py [hunk 6]
```diff
@@ -3313,6 +3316,78 @@ def test_parsing_errors(self):
...
+            pass
+
+        # Default strict=True, empty result expected
+        self.assertEqual(utils.getaddresses([address]), [empty])
+        self.assertEqual(utils.parseaddr([address]), empty)
+
+    @contextlib.contextmanager
+    def _email_strict_parsing_conf(self):
+        """Context for the given email strict parsing configured in config file"""
+        old_filename = utils._EMAIL_CONFIG_FILE
```
Comment (vstinner) [inline Lib/test/test_email/test_email.py]: You can use `test.support.swap_attr()` for that.

File: Lib/email/utils.py (hunk 3 above)
Comment (frenzymadness) [reply Lib/email/utils.py]: Do you think the current implementation is too complex so it needs another layer? Or is there any other benefit of the proposed solution I don't see?

File: Lib/test/test_email/test_email.py (hunk 4 above)
Comment (frenzymadness) [reply Lib/test/test_email/test_email.py:3321]: You are right but I consider this more like an end-to-end test.

File: Lib/test/test_email/test_email.py (hunk 5 above)
Comment (frenzymadness) [reply Lib/test/test_email/test_email.py:3329]: Sure thing.

File: Lib/email/utils.py (hunk 3 above)
Comment (vstinner) [reply Lib/email/utils.py]: It's just a coding style suggestion, IMO it makes the code easier to read/maintain. It's up to you.

File: Lib/test/test_email/test_email.py (hunk 6 above)
Comment (frenzymadness) [reply Lib/test/test_email/test_email.py]: Great idea, implemented in the last fixup commit.

File: Lib/test/test_email/test_email.py (hunk 5 above)
Comment (frenzymadness) [reply Lib/test/test_email/test_email.py:3329]: Implemented in the first fixup commit.

**Review by vstinner — APPROVED** [review]
LGTM. My remaining comments are just suggestions, not requirements.

**Comment by frenzymadness** [comment]
The last two fixup commits contain improved implementation as suggested by Victor and adjusted documentation for both functions.

Victor is on PTO. @torsava could you please take a look?

File: Lib/email/utils.py (hunk 3 above)
Comment (torsava) [reply Lib/email/utils.py]: I agree, I like the new implementation - nice separation of concerns.

**Comment by torsava** [comment]
I just reviewed it, and I think it looks very good, and all previous review comments were appropriately addressed. LGTM!

**Comment by frenzymadness** [comment]
Implemented in many places in RHEL.
